[CoordinatedGraphics] Do not unite dirty rects in Canvas2D

zhani opened the pull request:
#### b2410e35eca0141d7ab7eefda84365616f4d6f80
<pre>
[CoordinatedGraphics] Do not unite dirty rects in Canvas2D
<a href="https://bugs.webkit.org/show_bug.cgi?id=265365">https://bugs.webkit.org/show_bug.cgi?id=265365</a>

Reviewed by Carlos Garcia Campos.

CanvasRenderingContext2DBase tracks one dirty rect and does union of all
incoming dirty rects within repaint cycle. This cause unnecessary
invalidations of tiles in coordinated graphics. For example in case where
canvas spans over multiple tiles and only small areas at top-left and
bottom-right corners of canvas are changed. This scenario causes whole canvas
and all tiles to be marked dirty because of unification of dirty areas although
only top-left and bottom-right tiles are actually dirty.
 Solution is not to do unification of dirty rects in
CanvasRenderingContext2DBase. In coordinated graphics tiles do dirty rect
unification internally within tile.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::didDraw):

Canonical link: <a href="https://commits.webkit.org/271139@main">https://commits.webkit.org/271139@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39e12162247a15438cc0da986cafd6bf3ead6cbe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28715 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25143 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27935 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8021 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24936 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4894 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23581 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4263 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4420 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24577 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30329 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25001 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30553 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4442 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2578 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28526 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5908 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4899 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3549 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4829 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/39e12162247a15438cc0da986cafd6bf3ead6cbe. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/20917

zhani committed:
[CoordinatedGraphics] Do not unite dirty rects in Canvas2D
https://bugs.webkit.org/show_bug.cgi?id=265365

Reviewed by Carlos Garcia Campos.

CanvasRenderingContext2DBase tracks one dirty rect and does union of all
incoming dirty rects within repaint cycle. This cause unnecessary
invalidations of tiles in coordinated graphics. For example in case where
canvas spans over multiple tiles and only small areas at top-left and
bottom-right corners of canvas are changed. This scenario causes whole canvas
and all tiles to be marked dirty because of unification of dirty areas although
only top-left and bottom-right tiles are actually dirty.
 Solution is not to do unification of dirty rects in
CanvasRenderingContext2DBase. In coordinated graphics tiles do dirty rect
unification internally within tile.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::didDraw):

Canonical link: https://commits.webkit.org/271139@main

webkit-commit-queue commented:
Committed 271139@main (b2410e35eca0): <https://commits.webkit.org/271139@main>

Reviewed commits have been landed. Closing PR #20917 and removing active labels.
